### PR TITLE
Tensor observables

### PR DESCRIPTION
**Pull request by quantshah**
**Description of the Change:**

This PR will discuss how to implement a tensor of observables in Pennylane and implement it for the default plugins. Also, update the plugin documentation to include the possibility of tensored observables in hardware.

**Benefits:**
A nice feature to have perhaps for some algorithms.

**Possible Drawbacks:**
Maybe many hardware implementations do not have the possibility to perform measurement of tensored observables.

**Related GitHub Issues:**
#212

### CODE REVIEW

**Comment by josh146** [comment]
>  A hacky way to do this quick now would be to convert a list of observables into a multi-qubit Hermitian observable and then replacing it in the queue.

I would suggest that this logic (converting a list of tensor observables to a multi-qubit Hermitian matrix) should be left to the plugins - all the QNode should do is add to the queue an object that 'represents' the tensor of contained operations, and pass this directly to the device.

The reason being that the logic of how to apply tensor observables may differ depending on the plugin framework:

* `default.qubit`: the device will likely have to make use of the process you describe above, i.e., `DefaultQubitDevice` will recieve the abstract tensor operation, and manually perform `np.kron()` on the constituent single qubit hermitian matrix representations.

* `forrest.qpu`: pyQuil has built-in support for Pauli matrix algebra. The device could make  use of [operator estimation](http://docs.rigetti.com/en/stable/apidocs/autogen/pyquil.operator_estimation.measure_observables.html#pyquil.operator_estimation.measure_observables). However, this is likely a long term solution; in the short term, it will likely implement dense kronecker products as in default qubit.

* `qiskit` also has some methods for working with [tensor products of Pauli matrices](https://qiskit.org/documentation/autodoc/qiskit.quantum_info.operators.pauli.html), but I am not so familiar with them.

---

In terms of representing the tensor product, I'm not really sure what the best approach could be (it will likely change as you play around with the code), but one ideas could be:

If the list of arguments to `expval` is greater than one, i.e., `expval(PauliX(0), PauliY(1))`, the `expval` function could return a `Tensor` class, or maybe just a `Tensor` [named tuple](https://docs.python.org/3/library/collections.html#collections.namedtuple). The plugins would then have to be modified to deal with the new `Tensor` observable container.

**Comment by co9olguy** [comment]
Good point @josh146. You guys can ignore my comment in the related issue #212

**Comment by co9olguy** [comment]
We want to make things as easy as possible for plugin development, so the object that appears in the measurement queue should be pretty simple. Since measurement operators will know which wire they apply to, is it not enough to just pass a tuple of measurement operators to the backends, e.g., `(PauliZ(0), PauliY(1))`. Perhaps a tiny wrapper class for easier bookkeeping

**Comment by johannesjmeyer** [comment]
A thought:

In principle, we could measure multiple Tensor observables on multiple wires simultaneously (e.g. Z x Z and X x X on two wires). But this would require some serious changes to the codebase because one would have to check if they commute. Until this is realized, every plugin can just measure the individual operators of the tensor and multiply the outcomes to get the outcome for the tensor observable.

**Comment by quantshah** [comment]
Hi, this is ready for review. I have blacked it with -l 100 but still there are some parts which are reformatted by black in the files that I worked on. @co9olguy @josh146 

I will open a new issue to discuss if we should be passing just `obs.name, obs.wires, obs.parameter` `expval` or directly an `obs` object. This is relevant since for the `Tensor` object it is nice if we pass the object itself and `expval` takes care of it. Right now I do not see any advantage in passing strings over passing an `Operation (Observable)` to expval.

File: pennylane/_device.py [hunk 1]
```diff
@@ -222,14 +226,22 @@ def execute(self, queue, observables, parameters={}):
 
             for obs in observables:
                 if obs.return_type == "expectation":
-                    results.append(self.expval(obs.name, obs.wires, obs.parameters))
+                    if obs.name == "Tensor":
+                        # Pass the full tensor object rather than just names
+                        results.append(self.expval(obs, obs.wires, obs.parameters))
+                    else:
+                        results.append(self.expval(obs.name, obs.wires, obs.parameters))
```
Comment (josh146) [inline pennylane/_device.py]: @quantshah, I have a suggestion here that will avoid the `if` statement, and treat a `Tensor` object like any other observable.

For example, consider the following tensor:
```python
obs = Tensor()
obs += PauliX(wires=0)
obs += PauliZ(wires=1)
obs += Hermitian(A, wires=[2, 3])
```
With each observable appended, we could just have the class do
```python
    def append(self, op):
        """Add operations to a tensor list"""
        self.ops.append(op)
        self.name.append(op.name)
        self.wires.append(op.wires)
        self.params.append(op.params)

    __iadd__ = append
```

then we would have
```python
>>> obs.name
['PauliX', 'PauliZ', 'Hermitian']
>>> obs.wires
[[0], [1], [2, 3]]
>>> obs.parameters
[[], [], [A]]
```

It would then be up to the plugin to recognize that if `isinstance(obs.name, list)`, then the operation is a tensor (this could be part of the developer API).

File: pennylane/_device.py [hunk 2]
```diff
@@ -373,11 +394,22 @@ def check_validity(self, queue, observables):
...
+                raise DeviceError(
+                    "Observable {} not supported on device {}".format(o.name, self.short_name)
+                )
+
+            if isinstance(o, Tensor) and (self.capabilities()["tensor_observables"] is False):
+                raise DeviceError(
+                    "Tensor observables {} not supported on device {}".format(
+                        o.ops, self.short_name
+                    )
+                )
```
Comment (josh146) [inline pennylane/_device.py]: Nice! We should start to make proper use of the capabilities dictionary. Other things we should declare are

* `formulation` (qubit or CV)
* `hardware` (true or false)
* `state_accessible` (true or false, for simulators that allow the state vector to be extracted)

File: pennylane/operation.py [hunk 3]
```diff
@@ -695,5 +742,32 @@ def heisenberg_obs(self, num_wires):
...
+        self.wires = []
+        self.name = "Tensor"
+        self.parameters = None
+
+    def append(self, op):
+        """Add operations to a tensor list
+        """
+        self.ops.append(op)
+        for w in op.wires:
+            self.wires.append(w)
```
Comment (josh146) [inline pennylane/operation.py]: I just had a thought - do we need a separate Tensor class? Perhaps its functionality could be rolled directly into the `Observable` class, by adding the method
```python
    def append(self, op):
        """Add operations to a tensor list"""
        self.ops.append(op)
        self.name.append(op.name)
        self.wires.append(op.wires)
        self.params.append(op.params)
        return self

    __iadd__ = append
    __add__ = append
```

This way, to define a Tensor operation, you could simply do
```python
tensor_obs = PauliX(0) + Hermitian(A, wires=[1,2])
```
internally within the `expval`, `var`, or `sample` functions:
```python
if len(ops) > 1:
    # tensor of observables
    tensor_ops = ops[0]
    for o in ops:
        tensor_ops += o
    tensor_ops.return_type = "expectation"
````

Or perhaps I'm missing something that would stop this from working?

File: pennylane/plugins/default_qubit.py [hunk 4]
```diff
@@ -395,15 +454,22 @@ def mat_vec_product(self, mat, vec, wires):
...
         return np.reshape(state_multi_index, 2 ** self.num_wires)
 
     def expval(self, observable, wires, par):
         if self.shots == 0:
             # exact expectation value
-            A = self._get_operator_matrix(observable, par)
-            ev = self.ev(A, wires)
+            if isinstance(observable, Tensor):
+                # unpack the observable matrix
+                # TODO: merge the matrix generator functions
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]: What do you mean here by this TODO? Merge `_get_tensor_operator_matrix` with `_get_operator_matrix`?

File: pennylane/plugins/default_qubit.py [hunk 5]
```diff
@@ -395,15 +454,22 @@ def mat_vec_product(self, mat, vec, wires):
...
+                A = self._get_tensor_operator_matrix(observable)
+                ev = self.ev(A, wires)
+            else:
+                A = self._get_operator_matrix(observable, par)
+                ev = self.ev(A, wires)
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]: This line is duplicated by both branches of the if statement, so can be moved below the `if-else`.

File: pennylane/plugins/default_qubit.py [hunk 6]
```diff
@@ -453,6 +519,18 @@ def _get_operator_matrix(self, operation, par):
...
+        """Get the operator matrix for a given tensor of operations.
+
+        Args:
+          tensor (:class:`Tensor`): A tensor of operations
+          par (tuple[float]): parameter values
+        Returns:
+          array: matrix representation.
+        """
+        ops = [self._get_operator_matrix(op.name, op.parameters) for op in tensor.ops]
+        return np.kron(*ops)
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]: :+1:

File: pennylane/qnode.py [hunk 7]
```diff
@@ -999,34 +1051,37 @@ def _pd_analytic_var(self, param_values, param_idx, **kwargs):
...
+            ) from None
 
         return TFEQNode(self)
 
 
-#def QNode_vjp(ans, self, params, *args, **kwargs):
+# def QNode_vjp(ans, self, params, *args, **kwargs):
 def QNode_vjp(ans, self, args, **kwargs):
     """Returns the vector Jacobian product operator for a QNode, as a function
     of the QNode evaluation for specific argnums at the specified parameter values.
```
Comment (josh146) [inline pennylane/qnode.py]: Just double checking, but have you made any changes to `qnode.py` apart from blacking it? The formatting changes make it hard to find.

If no other changes have been made, perhaps replace this file with the one from `master`, to avoid it showing up in the diff.

File: tests/test_measure.py [hunk 8]
```diff
@@ -64,11 +64,52 @@ def circuit():
...
+            qml.RX(0.52, wires=0)
+            qml.RZ(0.52, wires=1)
+
+            return qml.expval(
+                qml.Hermitian(observable_matrix, wires=[0, 1]), qml.PauliY(2)
+            )
+
+        with pytest.raises(
+            QuantumFunctionError, match="Only single wire observables can be tensored."
+        ):
```
Comment (josh146) [inline tests/test_measure.py]: Maybe I missed something, but why is this the case?

Comment (josh146) [reply tests/test_measure.py]: Perhaps I'm wrong, but it should be easy to extend the tensor operation to `qml.Hermitian`

File: pennylane/operation.py (hunk 3 above)
Comment (co9olguy) [reply pennylane/operation.py]: @josh146 maybe from a user-interface perspective that's not to best way to do things? If there was a clear operator we could overload, then it makes sense, but to me `+` immediately brings to mind something like
X_0\otimes\Identity_1 + Hermitian_{0,1}

Comment (co9olguy) [reply pennylane/operation.py]: Alternate suggestions: 
```PauliX(0) @ Hermitian(A, wires=[0,1])``` <-- this one seems most intuitive
```PauliX(0) % Hermitian(A, wires=[0,1])```

Comment (josh146) [reply pennylane/operation.py]: Note that this won't be user facing, just used internally (so perhaps we shouldn't be operator overloading at all and just using the method 😆)

Instead, the user will be using the `expval(PauliX(0), PauliY(1))` syntax.

Comment (co9olguy) [reply pennylane/operation.py]: If it's useful, why not make it user-facing?

```
big_obs = PauliX(0) @ PauliY(1) @ PauliZ(2)
def circuit1(pars):
   ...
   return qml.expval(big_obs)

def circuit1(pars):
   ...
   return qml.sample(big_obs)
```
etc.

Comment (josh146) [reply pennylane/operation.py]: I don't mind that! it would have to be done from within the QNode though so that queueing can take place:

```python
def circuit1(pars):
   big_obs = PauliX(0) @ PauliY(1) @ PauliZ(2)
   return qml.expval(big_obs)
```

(More reason to try and solve/replace queuing.)

My only reason to slightly preferring `expval(PauliX(0), PauliY(1))` over `expval(PauliX(0) @ PauliY(1))` is just code complexity. The latter simply requires `*args` in the expval signature, whereas operator overloading requires new methods and could result in more edge-cases to test.

File: pennylane/operation.py [hunk 9]
```diff
@@ -695,5 +742,32 @@ def heisenberg_obs(self, num_wires):
...
         p = self.parameters
-        U = self._heisenberg_rep(p) # pylint: disable=assignment-from-none
+        U = self._heisenberg_rep(p)  # pylint: disable=assignment-from-none
         return self.heisenberg_expand(U, num_wires)
+
+
+class Tensor:
+    """Tensor wrapper for operations and observables.
+
+    :class:`Tensor` is used to wrap tensors of observables.
```
Comment (co9olguy) [inline pennylane/operation.py]: This part of the docstring seems redundant

File: pennylane/operation.py [hunk 10]
```diff
@@ -695,5 +742,32 @@ def heisenberg_obs(self, num_wires):
...
+
+    Args:
+        ops (list of Observable): Quantum observable or a list of observables.
```
Comment (co9olguy) [inline pennylane/operation.py]:
```suggestion
        ops (list[Observable]): Quantum observable or a list of observables.
```

File: tests/test_measure.py [hunk 11]
```diff
@@ -64,11 +64,52 @@ def circuit():
...
+
+    def test_tensored_observables(self):
+        """Test expval calculations for a tensor of observables."""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliY(wires=[0]), qml.PauliY(1))
```
Comment (co9olguy) [inline tests/test_measure.py]:
```suggestion
            return qml.expval(qml.PauliY(0), qml.PauliY(1))
```

File: tests/test_measure.py [hunk 12]
```diff
@@ -64,11 +64,52 @@ def circuit():
...
+
+        res = circuit()
+        assert np.allclose(res, -1.0)
```
Comment (co9olguy) [inline tests/test_measure.py]: Cool :)

File: pennylane/plugins/default_qubit.py [hunk 13]
```diff
@@ -453,6 +519,18 @@ def _get_operator_matrix(self, operation, par):
...
         return A(*par)
 
+    def _get_tensor_operator_matrix(self, tensor):
+        """Get the operator matrix for a given tensor of operations.
+
+        Args:
+          tensor (:class:`Tensor`): A tensor of operations
+          par (tuple[float]): parameter values
+        Returns:
+          array: matrix representation.
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]:
```suggestion
          array: matrix representation
```

File: pennylane/qnode.py (hunk 7 above)
Comment (co9olguy) [reply pennylane/qnode.py]: Looks like the changes were inserted at line 371

File: pennylane/plugins/default_qubit.py (hunk 6 above)
Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: If I'm reading right, this will always measure by passing the full dense matrix. @josh146 is there a case where it would make sense to pass the individual matrices of the tensor product to backend, without combining them?

File: pennylane/measure.py [hunk 14]
```diff
@@ -61,34 +61,64 @@ def circuit(x, y):
...
+            if QNode._current_context is not None:
+                # add observable to QNode observable queue
+                QNode._current_context._append_op(op)
+
+            return op
+
+        if len(ops) > 1:
+            # tensor of observables
+            tensor_ops = Tensor()
+            tensor_ops.return_type = "expectation"
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/measure.py]: Very easy from the plugin side :slightly_smiling_face:

File: pennylane/measure.py [hunk 15]
```diff
@@ -61,34 +61,64 @@ def circuit(x, y):
...
 
-        return op
+            for op in ops:
+                if len(op.wires) > 1:
+                    raise QuantumFunctionError("Only single wire observables can be tensored.")
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/measure.py]: That should also be easy, so should probably be included in this PR as well

File: pennylane/plugins/default_qubit.py (hunk 6 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: Currently, all PL does is pass the _names_ of the observable in the tensor product to the plugin, not even the matrices, so it's completely up to the plugin how it is implemented. I think this is the best approach due to the differences across frameworks.

For this particular case here, there are more efficient approaches we can make use of inside `default.qubit`. This could include performing single-qubit basis rotations, and simply measuring in the Z^{\otimes N} basis.

File: pennylane/operation.py (hunk 3 above)
Comment (johannesjmeyer) [reply pennylane/operation.py]: Even if this is a little bit old now (mabye you agreed on something in the meantime), I'd like to add that I really don't like `@` as an operator for tensor products because it clearly has the meaning of the matrix product in python and measuring `X @ H` could be mistaken as measuring the product of X and H and not the tensor product.

What would be possible is to construct [infix operators](http://tomerfiliba.com/blog/Infix-Operators/):
```python
qml.expval(qml.PauliX(0) |tensor| qml.PauliY(1) |tensor| qml.Hermitian(A, [3, 4]))
qml.expval(qml.PauliX(0) |x| qml.PauliY(1) |x| qml.Hermitian(A, [3, 4]))
qml.expval(qml.PauliX(0) <<x>> qml.PauliY(1) <<x>> qml.Hermitian(A, [3, 4]))
```
But I don't thinkt that would be a good idea because this will introduce a lot of complexity without much improvement over `Tensor(qml.PauliX(0), qml.PauliY(1), qml.Hermitian(A, [3, 4]))`

**Comment by josh146** [comment]
@co9olguy, I believe this is in a ready to review state now.

File: pennylane/measure.py [hunk 16]
```diff
@@ -57,70 +57,37 @@ def circuit(x, y):
...
+    if QNode._current_context is not None:
+        # delete operations from QNode queue
+        if isinstance(op, Tensor):
+            for o in op.obs:
+                QNode._current_context.queue.remove(o)
+        else:
+            QNode._current_context.queue.remove(op)
 
+    # set return type to be an expectation value
+    op.return_type = Expectation
```
Comment (co9olguy) [inline pennylane/measure.py]: is it strictly necessary to pull ops out of the queue just to update `return_type`, then put them right back?

File: pennylane/measure.py [hunk 17]
```diff
@@ -136,9 +103,13 @@ def var(op):
...
-        QNode._current_context.queue.remove(op)
+        if isinstance(op, Tensor):
+            for o in op.obs:
+                QNode._current_context.queue.remove(o)
+        else:
+            QNode._current_context.queue.remove(op)
 
-    # set return type to be a variance
+    # set return type to be an expectation value
     op.return_type = Variance
```
Comment (co9olguy) [inline pennylane/measure.py]: code doesn't seem to match comment

File: pennylane/measure.py [hunk 18]
```diff
@@ -161,14 +132,18 @@ def sample(op):
...
+        # delete operations from QNode queue
+        if isinstance(op, Tensor):
+            for o in op.obs:
+                QNode._current_context.queue.remove(o)
+        else:
+            QNode._current_context.queue.remove(op)
 
-    # set return type to be a sample
+    # set return type to be an expectation value
     op.return_type = Sample
```
Comment (co9olguy) [inline pennylane/measure.py]: same as above

File: pennylane/operation.py [hunk 19]
```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
...
+            else:
+                raise ValueError("Can only perform tensor products between observables.")
+
+    def __str__(self):
+        """Print the tensor product and some information."""
+        return 'Tensor product {}: {} params, wires {}'.format([i.name for i in self.obs], len(self.params), self.wires)
+
+    @property
+    def name(self):
+        """All consituent observable names making up the tensor product.
```
Comment (co9olguy) [inline pennylane/operation.py]:
```suggestion
        """All constituent observable names making up the tensor product.
```

File: pennylane/operation.py [hunk 20]
```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
...
+                self.obs.append(o)
+            else:
+                raise ValueError("Can only perform tensor products between observables.")
+
+    def __str__(self):
+        """Print the tensor product and some information."""
+        return 'Tensor product {}: {} params, wires {}'.format([i.name for i in self.obs], len(self.params), self.wires)
+
+    @property
+    def name(self):
```
Comment (co9olguy) [inline pennylane/operation.py]: We using (non-plural) `name` to keep code compatability?

File: pennylane/operation.py [hunk 21]
```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
...
+        return [o.name for o in self.obs]
+
+    @property
+    def num_wires(self):
+        """Number of wires the tensor product acts on.
+
+        Returns:
+            int: number of wires
+        """
+        return len([w for sublist in [o.wires for o in self.obs] for w in sublist])
```
Comment (co9olguy) [inline pennylane/operation.py]: `len(_flatten(self.wires))`?

File: pennylane/operation.py [hunk 22]
```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
...
+        return [o.wires for o in self.obs]
+
+    @property
+    def params(self):
+        """Raw parameters of all constituent observables in the tensor product.
+
+        Returns:
+            list[Any]: flattened list containing all dependent parameters
+        """
+        return [p for sublist in [o.params for o in self.obs] for p in sublist]
```
Comment (co9olguy) [inline pennylane/operation.py]: Can we not use `self.parameters` here?
What is the difference between `o.parameters` (used in `def parameters` below) and `o.params` used here? From what I can see, attempting to access `o.params` will raise an exception

Comment (co9olguy) [reply pennylane/operation.py]: Actually, I find having two functions `self.params` and `self.parameters` to be confusing

File: pennylane/operation.py [hunk 23]
```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
...
+        if isinstance(other, Observable):
+            self.obs.append(other)
+            return self
+
+        raise ValueError("Can only perform tensor products between observables.")
+
+    def __rmatmul__(self, other):
+        if isinstance(other, Observable):
+            self.obs[:0] = [other]
+            return self
```
Comment (co9olguy) [inline pennylane/operation.py]: I guess ordering ambiguities (`X @ Y` vs `Y @ X`) are resolved by the `wires` argument, so this is just for aesthetics?

File: pennylane/plugins/default_qubit.py [hunk 24]
```diff
@@ -470,16 +484,29 @@ def _get_operator_matrix(self, operation, par):
             return A
         return A(*par)
 
+    def _get_tensor_operator_matrix(self, obs, par):
+        """Get the operator matrix for a given tensor of operations.
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]:
```suggestion
        """Get the operator matrix for a given tensor product of operations.
```

File: pennylane/plugins/default_qubit.py [hunk 25]
```diff
@@ -470,16 +484,29 @@ def _get_operator_matrix(self, operation, par):
...
+
+        Args:
+            obs (list[str]): list of observable names to tensor
+            par (list[list[Any]]): parameter values
+
+        Returns:
+            array: matrix representation.
+        """
+        ops = [self._get_operator_matrix(o, p) for o, p in zip(obs, par)]
+        return functools.reduce(np.kron, ops)
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]: cool, how come you went with `reduce`?

File: pennylane/measure.py (hunk 16 above)
Comment (josh146) [reply pennylane/measure.py]: I agree, I really don't like this (it feels hacky). I guess there are three options, which differ in scope:

1. Re-think the queuing behaviour across the board.

2. Remove queuing from just observables (since it is not needed), and have the `expval/sample/var` functions perform the queuing

3. See if it's possible to find and update the observable in the queue, without removing it?

Comment (josh146) [reply pennylane/measure.py]: (We should make an issue for this)

File: pennylane/measure.py [hunk 26]
```diff
@@ -136,9 +103,13 @@ def var(op):
...
         # delete operations from QNode queue
-        QNode._current_context.queue.remove(op)
+        if isinstance(op, Tensor):
+            for o in op.obs:
+                QNode._current_context.queue.remove(o)
+        else:
+            QNode._current_context.queue.remove(op)
 
-    # set return type to be a variance
+    # set return type to be an expectation value
```
Comment (josh146) [inline pennylane/measure.py]:
```suggestion
    # set return type to be a variance
```

File: pennylane/measure.py (hunk 17 above)
Comment (josh146) [reply pennylane/measure.py]: A rogue copy and paste :(

File: pennylane/measure.py [hunk 27]
```diff
@@ -161,14 +132,18 @@ def sample(op):
...
-        QNode._current_context.queue.remove(op)
+        # delete operations from QNode queue
+        if isinstance(op, Tensor):
+            for o in op.obs:
+                QNode._current_context.queue.remove(o)
+        else:
+            QNode._current_context.queue.remove(op)
 
-    # set return type to be a sample
+    # set return type to be an expectation value
```
Comment (josh146) [inline pennylane/measure.py]:
```suggestion
    # set return type to be a sample
```

File: pennylane/operation.py (hunk 20 above)
Comment (josh146) [reply pennylane/operation.py]: Yes, unfortunately. Compatibility was not fun in this PR, and the main reason it took so long!

File: pennylane/operation.py (hunk 21 above)
Comment (josh146) [reply pennylane/operation.py]: Ah good point

Comment (josh146) [reply pennylane/operation.py]:
```suggestion
        return len(_flatten(self.wires))
```

File: pennylane/operation.py (hunk 22 above)
Comment (josh146) [reply pennylane/operation.py]: This is an artifact of the `Tensor` class needing to satisfy the `Operation` class API, and valid `Operations` need to have both defined :slightly_frowning_face:  Not ideal.

In the `Operation` class, `.params` is used for the unevaluated parameters (so could contain `Variable` instances), while `.parameters` are evaluated.

The QNode always queries `Operation.params`, while the device always queries `Operation.parameters`

File: pennylane/operation.py (hunk 23 above)
Comment (josh146) [reply pennylane/operation.py]: Yep. `__matmul__` takes into account left multiplication of a observable with something, this takes into account right multiplication. It's only an issue since `expval @ tensor` is possible.

File: pennylane/plugins/default_qubit.py (hunk 25 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: In my heart I'm a functional programmer :laughing: How would you have done it?

File: tests/test_device.py [hunk 28]
```diff
@@ -174,6 +174,45 @@ def test_check_validity_on_valid_queue(self, mock_device_supporting_paulis):
...
+
+        # mock device does not support Tensor product
+        with pytest.raises(DeviceError, match="Tensor observables not supported"):
+            mock_device_supporting_paulis.check_validity(queue, observables)
+
+    def test_check_validity_on_invalid_obserable_with_tensor_support(self, monkeypatch):
+        """Tests the function Device.check_validity with tensor support capability
+        but with an invalid observable"""
+        queue = [
+            qml.PauliX(wires=0, do_queue=False),
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_device.py]: Yes I agree!!!!! That's my preferred option.

[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_device.py]: It would be a very simple change as well, and wouldn't break any tests.

File: pennylane/operation.py (hunk 22 above)
Comment (co9olguy) [reply pennylane/operation.py]: Something for a quick refactor (name change would be sufficient, e.g., `evaluated_params`)?

File: tests/test_operation.py [hunk 29]
```diff
@@ -394,3 +393,148 @@ class DummyObserv(qml.operation.Observable):
...
+        X = qml.PauliX(0, do_queue=False)
+        Y = qml.CNOT(wires=[0, 1], do_queue=False)
+        Z = qml.PauliZ(0, do_queue=False)
+
+        with pytest.raises(ValueError, match="Can only perform tensor products between observables"):
+            X @ Y
+
+        with pytest.raises(ValueError, match="Can only perform tensor products between observables"):
+            T = X @ Z
+            T @ Y
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_operation.py]: I would like to support this, but I think it requires some thinking/planning. It would be nice to have a `TensorSum` class, that allows you to do something like

```
O = 2 * X @ Y + a Z @ H
```
where perhaps `a` is a free parameter. It's just a matter of working out how to let the QNode know the 'classical' gradient rule `dO/da = Z @ H`.

File: pennylane/plugins/default_qubit.py (hunk 25 above)
Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: :man_shrugging:

File: tests/test_tensor_measurements.py [hunk 30]
```diff
@@ -0,0 +1,467 @@
...
+                + np.sin(phi)
+                * (
+                    15
+                    + 8 * np.cos(varphi)
+                    - 11 * np.cos(2 * varphi)
+                    + 42 * np.sin(varphi)
+                    + 3 * np.sin(2 * varphi)
+                )
+            )
+        ) / 16
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_tensor_measurements.py]: This was Black though :thinking:

File: tests/test_tensor_measurements.py [hunk 31]
```diff
@@ -0,0 +1,467 @@
...
+
+        assert np.allclose(var, expected, atol=tol, rtol=0)
```
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_tensor_measurements.py]: :champagne:

File: tests/test_tensor_measurements.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply tests/test_tensor_measurements.py]: Doesn't mean it looks good :laughing:

File: tests/test_operation.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply tests/test_operation.py]: k we can leave for future consideration

File: tests/test_tensor_measurements.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (josh146) [reply tests/test_tensor_measurements.py]: Oops, I fixed this, not sure why it's not showing up as changed. This is what it looks like now:
Screenshot_20191009_131448

File: tests/test_device.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply tests/test_device.py]: I updated the relevant github issue #161

File: pennylane/measure.py (hunk 16 above)
Comment (co9olguy) [reply pennylane/measure.py]: agreed, perhaps repurpose #161

File: tests/test_operation.py [hunk 32]
```diff
@@ -15,11 +15,9 @@
 Unit tests for :mod:`pennylane.operation`.
 """
 import pytest
+import numpy as np
 
 import pennylane as qml
-import pennylane.ops
-import pennylane.operation as oo
-import pennylane.variable as ov
```
Comment (co9olguy) [inline tests/test_operation.py]: :laughing:

File: tests/test_tensor_measurements.py [hunk 33]
```diff
@@ -309,22 +309,9 @@ def circuit(a, b, c):
...
-                * (
-                    15
-                    + 8 * np.cos(varphi)
-                    - 11 * np.cos(2 * varphi)
-                    + 42 * np.sin(varphi)
-                    + 3 * np.sin(2 * varphi)
-                )
+                4 * np.cos(phi)
+                * (4 + 8 * np.cos(varphi) + np.cos(2 * varphi) - (1 + 6 * np.cos(varphi)) * np.sin(varphi))
+                + np.sin(phi) * (15 + 8 * np.cos(varphi) - 11 * np.cos(2 * varphi) + 42 * np.sin(varphi) + 3 * np.sin(2 * varphi))
```
Comment (co9olguy) [inline tests/test_tensor_measurements.py]: much better

**Comment by co9olguy** [comment]
:fireworks:
